File: core/tax_engine.py
import copy
from datetime import datetime, date
from .currency import Rule115CurrencyConverter

class TaxEngine:
    """
    Dual-Calendar Realized Gain & FIFO Lot Engine for Indian Residents holding US Equities.
    - Post-Budget 2024 Rules (LTCG > 24 months @ 12.5%, STCG <= 24 months @ Slab).
    - Rule 115 SBI TT Buying Rate conversion.
    - Slices gains by Indian FY (Apr 1 - Mar 31) vs US CY (Jan 1 - Dec 31).
    """

    CURRENT_MARKET_PRICES = {
        "VOO": 515.20,
        "QQQ": 478.50,
        "QQQM": 196.40,
        "IVV": 516.80,
        "SPY": 514.90,
        "NVDA": 126.50,
        "AAPL": 228.40,
        "MSFT": 412.00,
        "AMD": 118.25,
        "TSLA": 210.50,
        "GOOGL": 182.10,
        "SMH": 242.00,
        "VGT": 545.00,
        "XLK": 215.00,
        "ARKK": 48.20,
    }

    def __init__(self, currency_converter=None):
        self.fx = currency_converter or Rule115CurrencyConverter()

    @staticmethod
    def get_indian_fy(date_str):
        """
        Returns Indian Financial Year string, e.g. 'FY 2024-25' for 2024-05-18.
        April 1 of Year N to March 31 of Year N+1.
        """
        dt = datetime.strptime(date_str[:10], "%Y-%m-%d").date()
        year = dt.year
        if dt.month >= 4:
            return f"FY {year}-{str(year + 1)[-2:]}"
        else:
            return f"FY {year - 1}-{str(year)[-2:]}"

    @staticmethod
    def get_us_cy(date_str):
        """
        Returns US Calendar Year string, e.g. 'CY 2024' for 2024-05-18.
        """
        dt = datetime.strptime(date_str[:10], "%Y-%m-%d").date()
        return f"CY {dt.year}"

    @staticmethod
    def is_post_budget_2024(date_str):
        """
        Budget 2024 changes took effect on July 23, 2024.
        """
        dt = datetime.strptime(date_str[:10], "%Y-%m-%d").date()
        return dt >= date(2024, 7, 23)

    def calculate(self, trades, tax_slab_pct=30.0, current_date="2026-03-25"):
        """
        Runs FIFO lot matching, calculates realized gains, dividends, open positions,
        and currency alpha decomposition.
        """
        # Separate trades
        buy_lots = {}  # symbol -> list of lots
        realized_trades = []
        dividends = []

        for trade in trades:
            action = trade["action"]
            symbol = trade["symbol"]

            if action == "BUY":
                if symbol not in buy_lots:
                    buy_lots[symbol] = []
                buy_fx = self.fx.get_rule_115_rate(trade["date"])
                buy_lots[symbol].append({
                    "trade_id": f"BUY-{trade['date']}-{symbol}-{len(buy_lots[symbol])+1}",
                    "date": trade["date"],
                    "symbol": symbol,
                    "asset_type": trade["asset_type"],
                    "total_quantity": trade["quantity"],
                    "remaining_quantity": trade["quantity"],
                    "price_usd": trade["price_usd"],
                    "fees_usd": trade["fees_usd"],
                    "buy_fx": buy_fx,
                    "cost_basis_usd": trade["quantity"] * trade["price_usd"] + trade["fees_usd"],
                    "cost_basis_inr": (trade["quantity"] * trade["price_usd"] + trade["fees_usd"]) * buy_fx,
                    "notes": trade.get("notes", "")
                })

            elif action == "SELL":
                sell_qty_needed = trade["quantity"]
                sell_price_usd = trade["price_usd"]
                sell_fees_usd = trade["fees_usd"]
                sell_date = trade["date"]
                sell_fx = self.fx.get_rule_115_rate(sell_date)

                if symbol not in buy_lots or not buy_lots[symbol]:
                    # Unmatched sell (or short/vested prior)
                    continue

                for lot in buy_lots[symbol]:
                    if sell_qty_needed <= 0:
                        break
                    if lot["remaining_quantity"] <= 0:
                        continue

                    matched_qty = min(lot["remaining_quantity"], sell_qty_needed)
                    lot["remaining_quantity"] -= matched_qty
                    sell_qty_needed -= matched_qty

                    # Fraction of fees
                    matched_buy_fees = (matched_qty / lot["total_quantity"]) * lot["fees_usd"] if lot["total_quantity"] > 0 else 0
                    matched_sell_fees = (matched_qty / trade["quantity"]) * sell_fees_usd if trade["quantity"] > 0 else 0

                    buy_cost_usd = (matched_qty * lot["price_usd"]) + matched_buy_fees
                    sell_proceeds_usd = (matched_qty * sell_price_usd) - matched_sell_fees

                    buy_cost_inr = buy_cost_usd * lot["buy_fx"]
                    sell_proceeds_inr = sell_proceeds_usd * sell_fx

                    realized_gain_usd = sell_proceeds_usd - buy_cost_usd
                    realized_gain_inr = sell_proceeds_inr - buy_cost_inr

                    # Holding period
                    dt_buy = datetime.strptime(lot["date"], "%Y-%m-%d").date()
                    dt_sell = datetime.strptime(sell_date, "%Y-%m-%d").date()
                    holding_days = (dt_sell - dt_buy).days
                    holding_months = round(holding_days / 30.4375, 1)

                    # LTCG vs STCG classification
                    post_budget = self.is_post_budget_2024(sell_date)
                    threshold_months = 24.0 if post_budget else 36.0
                    is_ltcg = holding_months > threshold_months

                    tax_rate_pct = 12.5 if is_ltcg else tax_slab_pct
                    estimated_tax_inr = max(0.0, realized_gain_inr * (tax_rate_pct / 100.0))

                    # Decompose return into Asset vs FX Drift
                    decomp = self.fx.decompose_return(
                        lot["price_usd"] * matched_qty,
                        sell_price_usd * matched_qty,
                        lot["date"],
                        sell_date
                    )

                    realized_trades.append({
                        "symbol": symbol,
                        "asset_type": trade["asset_type"],
                        "quantity": round(matched_qty, 4),
                        "buy_date": lot["date"],
                        "sell_date": sell_date,
                        "holding_days": holding_days,
                        "holding_months": holding_months,
                        "term": "LTCG" if is_ltcg else "STCG",
                        "buy_price_usd": lot["price_usd"],
                        "sell_price_usd": sell_price_usd,
                        "buy_fx_rate": lot["buy_fx"],
                        "sell_fx_rate": sell_fx,
                        "buy_cost_usd": round(buy_cost_usd, 2),
                        "sell_proceeds_usd": round(sell_proceeds_usd, 2),
                        "gain_usd": round(realized_gain_usd, 2),
                        "buy_cost_inr": round(buy_cost_inr, 2),
                        "sell_proceeds_inr": round(sell_proceeds_inr, 2),
                        "gain_inr": round(realized_gain_inr, 2),
                        "tax_rate_pct": tax_rate_pct,
                        "estimated_tax_inr": round(estimated_tax_inr, 2),
                        "indian_fy": self.get_indian_fy(sell_date),
                        "us_cy": self.get_us_cy(sell_date),
                        "asset_gain_usd": decomp["asset_gain_usd"],
                        "fx_gain_inr": decomp["fx_gain_inr"],
                        "notes": trade.get("notes", "")
                    })

            elif action == "DIVIDEND":
                div_fx = self.fx.get_rule_115_rate(trade["date"])
                gross_div_usd = trade["quantity"] * trade["price_usd"]
                withholding_usd = trade["withholding_tax_usd"]
                net_div_usd = gross_div_usd - withholding_usd

                gross_div_inr = round(gross_div_usd * div_fx, 2)
                withholding_inr = round(withholding_usd * div_fx, 2)
                net_div_inr = round(net_div_usd * div_fx, 2)

                # DTAA Sec 90 allows credit up to Indian tax or US WHT
                indian_tax_on_div = round(gross_div_inr * (tax_slab_pct / 100.0), 2)
                ftc_eligible_inr = min(withholding_inr, indian_tax_on_div)
                net_tax_payable_inr = max(0.0, indian_tax_on_div - ftc_eligible_inr)

                dividends.append({
                    "date": trade["date"],
                    "symbol": symbol,
                    "asset_type": trade["asset_type"],
                    "quantity": trade["quantity"],
                    "per_share_usd": trade["price_usd"],
                    "gross_div_usd": round(gross_div_usd, 2),
                    "withholding_usd": round(withholding_usd, 2),
                    "net_div_usd": round(net_div_usd, 2),
                    "fx_rate": div_fx,
                    "gross_div_inr": gross_div_inr,
                    "withholding_inr": withholding_inr,
                    "net_div_inr": net_div_inr,
                    "ftc_eligible_inr": ftc_eligible_inr,
                    "indian_tax_inr": indian_tax_on_div,
                    "net_tax_payable_inr": net_tax_payable_inr,
                    "indian_fy": self.get_indian_fy(trade["date"]),
                    "us_cy": self.get_us_cy(trade["date"]),
                })

        # Calculate Open Positions
        open_positions = []
        current_fx = self.fx.get_rule_115_rate(current_date)
        dt_current = datetime.strptime(current_date[:10], "%Y-%m-%d").date()

        for symbol, lots in buy_lots.items():
            for lot in lots:
                rem_qty = lot["remaining_quantity"]
                if rem_qty <= 0:
                    continue

                curr_price_usd = self.CURRENT_MARKET_PRICES.get(symbol, lot["price_usd"] * 1.05)
                mkt_val_usd = rem_qty * curr_price_usd
                mkt_val_inr = mkt_val_usd * current_fx

                cost_basis_usd = rem_qty * lot["price_usd"]
                cost_basis_inr = cost_basis_usd * lot["buy_fx"]

                unrealized_pnl_usd = mkt_val_usd - cost_basis_usd
                unrealized_pnl_inr = mkt_val_inr - cost_basis_inr
                unrealized_pnl_pct = ((curr_price_usd / lot["price_usd"]) - 1.0) * 100.0 if lot["price_usd"] > 0 else 0.0

                dt_buy = datetime.strptime(lot["date"], "%Y-%m-%d").date()
                holding_days = (dt_current - dt_buy).days
                holding_months = round(holding_days / 30.4375, 1)
                is_ltcg = holding_months > 24.0

                decomp = self.fx.decompose_return(cost_basis_usd, mkt_val_usd, lot["date"], current_date)

                open_positions.append({
                    "lot_id": lot["trade_id"],
                    "symbol": symbol,
                    "asset_type": lot["asset_type"],
                    "buy_date": lot["date"],
                    "quantity": round(rem_qty, 4),
                    "buy_price_usd": lot["price_usd"],
                    "current_price_usd": round(curr_price_usd, 2),
                    "buy_fx_rate": lot["buy_fx"],
                    "current_fx_rate": current_fx,
                    "cost_basis_usd": round(cost_basis_usd, 2),
                    "current_val_usd": round(mkt_val_usd, 2),
                    "cost_basis_inr": round(cost_basis_inr, 2),
                    "current_val_inr": round(mkt_val_inr, 2),
                    "unrealized_pnl_usd": round(unrealized_pnl_usd, 2),
                    "unrealized_pnl_inr": round(unrealized_pnl_inr, 2),
                    "unrealized_pnl_pct": round(unrealized_pnl_pct, 2),
                    "holding_days": holding_days,
                    "holding_months": holding_months,
                    "term": "LTCG" if is_ltcg else "STCG",
                    "asset_growth_usd": decomp["asset_gain_usd"],
                    "fx_drift_inr": decomp["fx_gain_inr"],
                    "is_loss": unrealized_pnl_inr < 0,
                    "notes": lot.get("notes", "")
                })

        # Summarize by Indian FY
        fy_summary = {}
        for r in realized_trades:
            fy = r["indian_fy"]
            if fy not in fy_summary:
                fy_summary[fy] = {
                    "stcg_gain_inr": 0.0,
                    "ltcg_gain_inr": 0.0,
                    "total_gain_inr": 0.0,
                    "stcg_tax_inr": 0.0,
                    "ltcg_tax_inr": 0.0,
                    "total_tax_inr": 0.0,
                    "gain_usd": 0.0,
                    "trades_count": 0
                }
            if r["term"] == "STCG":
                fy_summary[fy]["stcg_gain_inr"] += r["gain_inr"]
                fy_summary[fy]["stcg_tax_inr"] += r["estimated_tax_inr"]
            else:
                fy_summary[fy]["ltcg_gain_inr"] += r["gain_inr"]
                fy_summary[fy]["ltcg_tax_inr"] += r["estimated_tax_inr"]

            fy_summary[fy]["total_gain_inr"] += r["gain_inr"]
            fy_summary[fy]["total_tax_inr"] += r["estimated_tax_inr"]
            fy_summary[fy]["gain_usd"] += r["gain_usd"]
            fy_summary[fy]["trades_count"] += 1

        # Round FY summaries
        for fy in fy_summary:
            for k in fy_summary[fy]:
                if isinstance(fy_summary[fy][k], float):
                    fy_summary[fy][k] = round(fy_summary[fy][k], 2)

        # Totals across all open positions
        total_open_usd = sum(p["current_val_usd"] for p in open_positions)
        total_open_inr = sum(p["current_val_inr"] for p in open_positions)
        total_unrealized_pnl_usd = sum(p["unrealized_pnl_usd"] for p in open_positions)
        total_unrealized_pnl_inr = sum(p["unrealized_pnl_inr"] for p in open_positions)

        return {
            "realized_trades": realized_trades,
            "dividends": dividends,
            "open_positions": open_positions,
            "fy_summary": fy_summary,
            "portfolio_totals": {
                "current_val_usd": round(total_open_usd, 2),
                "current_val_inr": round(total_open_inr, 2),
                "unrealized_pnl_usd": round(total_unrealized_pnl_usd, 2),
                "unrealized_pnl_inr": round(total_unrealized_pnl_inr, 2),
                "open_lots_count": len(open_positions),
                "current_fx_rate": current_fx
            }
        }

File: packages/engine-ts/src/tax_engine.ts
/**
 * Paasa TaxAlpha Engine: FIFO Lot Matching & Budget 2024 Engine (TypeScript)
 * --------------------------------------------------------------------------
 * Matches sell orders against historical buy lots using strict FIFO.
 * Categorizes gains under Section 2(42A) Budget 2024:
 * - Holding > 24 months = LTCG @ 12.5% (flat)
 * - Holding <= 24 months = STCG @ applicable slab (20%, 30%, 39%)
 */

import type { TradeRecord, MatchedFIFOTrade, OpenLotPosition, CapitalGainsTerm } from "./types.ts";
import { Rule115CurrencyConverter } from "./currency.ts";

export class FIFOTaxEngine {
  private fx: Rule115CurrencyConverter;

  constructor(customFx?: Rule115CurrencyConverter) {
    this.fx = customFx || new Rule115CurrencyConverter();
  }

  getIndianFinancialYear(dateStr: string): string {
    const d = new Date(dateStr.slice(0, 10));
    const year = d.getUTCFullYear();
    const month = d.getUTCMonth() + 1; // 1-12
    if (month >= 4) {
      return `FY ${year}-${(year + 1).toString().slice(2)}`;
    } else {
      return `FY ${year - 1}-${year.toString().slice(2)}`;
    }
  }

  calculate(trades: TradeRecord[], taxSlabPct: number = 30.0, currentMarketPrices?: Record<string, number>): {
    matchedTrades: MatchedFIFOTrade[];
    openPositions: OpenLotPosition[];
    totalGainInr: number;
    totalTaxInr: number;
    fySummaries: Record<string, { totalGainInr: number; totalTaxInr: number; tradesCount: number }>;
  } {
    const sorted = [...trades].sort((a, b) => a.date.localeCompare(b.date));
    const buyQueues: Record<string, Array<{ trade: TradeRecord; remQty: number }>> = {};
    const matched: MatchedFIFOTrade[] = [];

    const defaultPrices: Record<string, number> = {
      "VOO": 515.20, "QQQ": 478.50, "NVDA": 128.40, "AMD": 118.25,
      "TSLA": 210.50, "AAPL": 224.30, "MSFT": 448.50, "AMZN": 186.40,
      ...currentMarketPrices
    };

    for (const t of sorted) {
      if (t.action === "BUY") {
        if (!buyQueues[t.symbol]) buyQueues[t.symbol] = [];
        buyQueues[t.symbol].push({ trade: t, remQty: t.quantity });
      } else if (t.action === "SELL") {
        let sellRem = t.quantity;
        const queue = buyQueues[t.symbol] || [];

        while (sellRem > 0.00001 && queue.length > 0) {
          const buyLot = queue[0];
          const matchedQty = Math.min(sellRem, buyLot.remQty);

          const buyDt = new Date(buyLot.trade.date);
          const sellDt = new Date(t.date);
          const holdingDays = Math.max(1, Math.round((sellDt.getTime() - buyDt.getTime()) / (1000 * 60 * 60 * 24)));
          const holdingMonths = Math.round((holdingDays / 30.4375) * 10) / 10;
          const isLtcg = holdingMonths > 24.0;
          const term: CapitalGainsTerm = isLtcg ? "LTCG" : "STCG";

          const buyFx = this.fx.getRule115Rate(buyLot.trade.date);
          const sellFx = this.fx.getRule115Rate(t.date);

          const costBasisUsd = matchedQty * buyLot.trade.priceUsd;
          const proceedsUsd = matchedQty * t.priceUsd;
          const gainUsd = proceedsUsd - costBasisUsd;

          const costBasisInr = costBasisUsd * buyFx;
          const proceedsInr = proceedsUsd * sellFx;
          const gainInr = proceedsInr - costBasisInr;

          const decomp = this.fx.decomposeReturn(costBasisUsd, proceedsUsd, buyLot.trade.date, t.date);

          const taxRate = isLtcg ? 12.5 : taxSlabPct;
          const estimatedTaxInr = gainInr > 0 ? (gainInr * taxRate) / 100.0 : 0.0;

          matched.push({
            buyTradeId: buyLot.trade.tradeId,
            sellTradeId: t.tradeId,
            symbol: t.symbol,
            assetType: t.assetType,
            quantity: Math.round(matchedQty * 10000) / 10000,
            buyDate: buyLot.trade.date,
            sellDate: t.date,
            holdingDays,
            holdingMonths,
            term,
            costBasisUsd: Math.round(costBasisUsd * 100) / 100,
            proceedsUsd: Math.round(proceedsUsd * 100) / 100,
            gainUsd: Math.round(gainUsd * 100) / 100,
            buyFxRate: buyFx,
            sellFxRate: sellFx,
            costBasisInr: Math.round(costBasisInr * 100) / 100,
            proceedsInr: Math.round(proceedsInr * 100) / 100,
            gainInr: Math.round(gainInr * 100) / 100,
            assetGrowthUsd: decomp.assetGainUsd,
            currencyDriftInr: decomp.currencyGainInr,
            applicableTaxRatePct: taxRate,
            estimatedTaxInr: Math.round(estimatedTaxInr * 100) / 100,
            indianFy: this.getIndianFinancialYear(t.date)
          });

          buyLot.remQty -= matchedQty;
          sellRem -= matchedQty;

          if (buyLot.remQty <= 0.00001) {
            queue.shift();
          }
        }
      }
    }

    // Open Positions
    const openPositions: OpenLotPosition[] = [];
    const currentDateStr = new Date().toISOString().slice(0, 10);
    const currentFx = this.fx.getRule115Rate(currentDateStr);

    for (const [sym, queue] of Object.entries(buyQueues)) {
      for (const item of queue) {
        if (item.remQty <= 0.00001) continue;
        const currPrice = defaultPrices[sym] || item.trade.priceUsd * 1.05;
        const buyFx = this.fx.getRule115Rate(item.trade.date);

        const costUsd = item.remQty * item.trade.priceUsd;
        const currValUsd = item.remQty * currPrice;
        const costInr = costUsd * buyFx;
        const currValInr = currValUsd * currentFx;

        const pnlUsd = currValUsd - costUsd;
        const pnlInr = currValInr - costInr;
        const pnlPct = item.trade.priceUsd > 0 ? ((currPrice - item.trade.priceUsd) / item.trade.priceUsd) * 100 : 0;

        const buyDt = new Date(item.trade.date);
        const nowDt = new Date(currentDateStr);
        const holdingDays = Math.max(1, Math.round((nowDt.getTime() - buyDt.getTime()) / (1000 * 60 * 60 * 24)));
        const holdingMonths = Math.round((holdingDays / 30.4375) * 10) / 10;
        const term: CapitalGainsTerm = holdingMonths > 24.0 ? "LTCG" : "STCG";

        openPositions.push({
          lotId: item.trade.tradeId,
          symbol: sym,
          assetType: item.trade.assetType,
          buyDate: item.trade.date,
          quantity: Math.round(item.remQty * 10000) / 10000,
          buyPriceUsd: item.trade.priceUsd,
          currentPriceUsd: currPrice,
          buyFxRate: buyFx,
          currentFxRate: currentFx,
          costBasisUsd: Math.round(costUsd * 100) / 100,
          currentValUsd: Math.round(currValUsd * 100) / 100,
          costBasisInr: Math.round(costInr * 100) / 100,
          currentValInr: Math.round(currValInr * 100) / 100,
          unrealizedPnlUsd: Math.round(pnlUsd * 100) / 100,
          unrealizedPnlInr: Math.round(pnlInr * 100) / 100,
          unrealizedPnlPct: Math.round(pnlPct * 100) / 100,
          holdingDays,
          holdingMonths,
          term,
          isLoss: pnlInr < 0
        });
      }
    }

    // Rollups
    let totalGainInr = 0;
    let totalTaxInr = 0;
    const fySummaries: Record<string, { totalGainInr: number; totalTaxInr: number; tradesCount: number }> = {};

    for (const m of matched) {
      totalGainInr += m.gainInr;
      totalTaxInr += m.estimatedTaxInr;

      if (!fySummaries[m.indianFy]) {
        fySummaries[m.indianFy] = { totalGainInr: 0, totalTaxInr: 0, tradesCount: 0 };
      }
      fySummaries[m.indianFy].totalGainInr += m.gainInr;
      fySummaries[m.indianFy].totalTaxInr += m.estimatedTaxInr;
      fySummaries[m.indianFy].tradesCount += 1;
    }

    return {
      matchedTrades: matched,
      openPositions,
      totalGainInr: Math.round(totalGainInr * 100) / 100,
      totalTaxInr: Math.round(totalTaxInr * 100) / 100,
      fySummaries
    };
  }
}
